gensim/models/doc2vec.py

```diff
@@ -220,29 +542,113 @@ def _vocab_from(sentences):
+                       for sentence, (doctag_indices, doctag_vectors, doctag_locks, ignored) in job)
+        elif self.dm_concat:
+            tally = sum(train_sentence_dm_concat(self, sentence, doctag_indices, alpha, work, neu1,
+                                                doctag_vectors=doctag_vectors, doctag_locks=doctag_locks)
+                       for sentence, (doctag_indices, doctag_vectors, doctag_locks, ignored) in job)
         else:
-            return sum(train_sentence_dm(self, sentence, lbls, alpha, work, neu1, self.train_words, self.train_lbls) for sentence, lbls in job)
+            tally = sum(train_sentence_dm(self, sentence, doctag_indices, alpha, work, neu1,
+                                         doctag_vectors=doctag_vectors, doctag_locks=doctag_locks)
+                       for sentence, (doctag_indices, doctag_vectors, doctag_locks, ignored) in job)
+        self.docvecs.trained_items(item for s, item in job)
+        return tally
+
+    def infer_vector(self, document, alpha=0.1, min_alpha=0.0001, steps=5):
+        """
+        Infer a vector for given post-bulk training document.
+
+        Document should be a list of (word) tokens.
+        """
+        doctag_vectors = empty((1, self.vector_size), dtype=REAL)
+        doctag_vectors[0] = self.seeded_vector(' '.join(document))
+        doctag_locks = ones(1, dtype=REAL)
+        doctag_indices = [0]
+        word_vocabs = self._tokens_to_vocabs(document)
+
+        work = zeros(self.layer1_size, dtype=REAL)
+        if not self.sg:
+            neu1 = matutils.zeros_aligned(self.layer1_size, dtype=REAL)
+
+        for i in range(steps):
+            if self.sg:
+                train_sentence_dbow(self, word_vocabs, doctag_indices, alpha, work,
+                                    learn_words=False, learn_hidden=False,
+                                    doctag_vectors=doctag_vectors, doctag_locks=doctag_locks)
+            elif self.dm_concat:
+                train_sentence_dm_concat(self, word_vocabs, doctag_indices, alpha, work, neu1,
+                                         learn_words=False, learn_hidden=False,
+                                         doctag_vectors=doctag_vectors, doctag_locks=doctag_locks)
+            else:
+                train_sentence_dm(self, word_vocabs, doctag_indices, alpha, work, neu1,
+                                  learn_words=False, learn_hidden=False,
+                                  doctag_vectors=doctag_vectors, doctag_locks=doctag_locks)
+            alpha = ((alpha - min_alpha) / (steps - i)) + min_alpha
+
+        return doctag_vectors[0]
 
     def __str__(self):
-        return "Doc2Vec(vocab=%s, size=%s, alpha=%s)" % (len(self.index2word), self.layer1_size, self.alpha)
+        return "Doc2Vec(%id, sg=%i, hs=%i, negative=%i, dm_concat=%i)" % (self.vector_size, self.sg, self.hs, self.negative, self.dm_concat)
+
+    @property
+    def compact_name(self):
```

Comment:
Sure! (I'd then still lead with the class name and abbreviate all else inside the parens.)